// @sap-ux/preview-middleware/dist/client/cpe/odata-health/odata-health-checker.js
"use strict";

sap.ui.define(["sap/base/Log", "sap/ui/model/odata/v2/ODataModel", "sap/ui/model/odata/v4/ODataModel", "./odata-health-status"], function (Log, ODataModelV2, ODataModelV4, ___odata_health_status) {
  "use strict";

  const ODataDownStatus = ___odata_health_status["ODataDownStatus"];
  const ODataUpStatus = ___odata_health_status["ODataUpStatus"];
  /**
   * The OData version type.
   */
  /**
   * Describes an OData service instance.
   */
  /**
   * Use this class to do a health check for all available OData services, supports both v2 and v4
   * format. This health checker ensures not only that $metadata is valid, but also that the UI5 framework
   * itself can consume the service via its models.
   */
  class ODataHealthChecker {
    /**
     * The OData type.
     */
    static ODATA_TYPE = 'OData';

    /**
     * The default OData version.
     */
    static DEFAULT_ODATA_VERSION = 'v2';

    /**
     * Use this helper function to filter the OData data source items from the manifest.
     * @param src The service data source.
     * @returns True if the data source represents an OData service.
     */
    isOdataService = src => src.type === ODataHealthChecker.ODATA_TYPE;

    /**
     * Use this helper function to map the OData data source to the internal structure
     * used in this class.
     * @param src The OData service data source.
     * @returns The OData service info object.
     */
    toOdataServiceInfo = src => ({
      serviceUrl: src.uri,
      oDataVersion: src.settings?.odataVersion ?? ODataHealthChecker.DEFAULT_ODATA_VERSION
    });
    constructor(rta) {
      this.rta = rta;
    }

    /**
     * Does a health check to all available OData services.
     *
     * @returns Resolves with an array containing the health
     * status for each OData service.
     */
    async getHealthStatus() {
      const oDataHealthCheckStartTime = Date.now();
      const services = this.getServices();
      const metadataPromises = await Promise.allSettled(services.map(({
        serviceUrl,
        oDataVersion
      }) => this.getServiceMetadata(serviceUrl, oDataVersion)));
      const oDataHealthCheckDurationInSec = ((Date.now() - oDataHealthCheckStartTime) / 1000).toFixed(2);
      Log.info(`OData service health check took ${oDataHealthCheckDurationInSec} sec.`);
      return metadataPromises.map((metadataPromise, idx) => metadataPromise.status === 'fulfilled' ? new ODataUpStatus(services[idx].serviceUrl) : new ODataDownStatus(services[idx].serviceUrl, metadataPromise.reason));
    }

    /**
     * This method does a strong health check (with the ODataModel). This ensures not only
     * that the $metadata is valid, but also that the UI5 framework itself can consume the service via its models.
     * Some services may have valid $metadata but still fail in the UI5’s ODataModel
     * (e.g., weird annotations, CORS issues, etc.).
     *
     * @param serviceUrl The OData service url.
     * @param oDataVersion The OData version.
     * @returns Resolved with valid metadata.
     */
    getServiceMetadata(serviceUrl, oDataVersion) {
      switch (oDataVersion) {
        case 'v2':
        case '2.0':
          return this.getServiceV2Metadata(serviceUrl);
        case 'v4':
        case '4.0':
          return this.getServiceV4Metadata(serviceUrl);
      }
    }
    getServiceV2Metadata(serviceUrl) {
      const oModel = new ODataModelV2({
        serviceUrl,
        json: true,
        // We do not want the annotations concatenated to the final result.
        loadAnnotationsJoined: false
      });
      // This method actually returns promise which is resolved with the metadata.
      return oModel.metadataLoaded(true).finally(
      // Do clean up in case the helath check is done multiple times.
      () => oModel.destroy());
    }
    getServiceV4Metadata(serviceUrl) {
      const oModel = new ODataModelV4({
        serviceUrl,
        // Only metadata loaded. We only want the model to load $metadata,
        // not fetch entity data or bind to any UI controls.
        synchronizationMode: 'None'
      });
      // This method actually returns promise which is resolved with the metadata.
      return oModel.getMetaModel().requestObject('/').finally(
      // Do clean up in case the helath check is done multiple times.
      () => oModel.destroy());
    }
    getServices() {
      const manifest = this.rta.getRootControlInstance().getManifest();
      const dataSources = manifest?.['sap.app']?.dataSources;
      return Object.values(dataSources ?? {}).filter(this.isOdataService).map(this.toOdataServiceInfo);
    }
  }
  var __exports = {
    __esModule: true
  };
  __exports.ODataHealthChecker = ODataHealthChecker;
  return __exports;
});
