// @sap-ux/adp-tooling/dist/writer/project-utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getPackageJSONInfo = getPackageJSONInfo;
exports.getTypes = getTypes;
exports.getCfVariant = getCfVariant;
exports.writeTemplateToFolder = writeTemplateToFolder;
exports.writeUI5Yaml = writeUI5Yaml;
exports.writeCfUI5Yaml = writeCfUI5Yaml;
exports.writeUI5DeployYaml = writeUI5DeployYaml;
exports.writeCfTemplates = writeCfTemplates;
const node_path_1 = require("node:path");
const node_fs_1 = require("node:fs");
const options_1 = require("./options");
const ui5_config_1 = require("@sap-ux/ui5-config");
/**
 * Retrieves the package name and version from the package.json file located two levels up the directory tree.
 *
 * @returns {Package} An object containing the `name` and `version` of the package.
 */
function getPackageJSONInfo() {
    const defaultPackage = {
        name: '@sap-ux/adp-tooling',
        version: 'NO_VERSION_FOUND'
    };
    try {
        return JSON.parse((0, node_fs_1.readFileSync)((0, node_path_1.join)(__dirname, '../../package.json'), 'utf-8'));
    }
    catch (e) {
        return defaultPackage;
    }
}
/**
 * Determines the correct TypeScript definitions package and version based on a given UI5 version.
 *
 * If the version includes `"snapshot"`, it returns a predefined default types package and version.
 * Otherwise, it selects the appropriate package and computes the corresponding version using either
 * `getTypesVersion` or `getEsmTypesVersion`.
 *
 * @param {string} [ui5Version] - The version of UI5 (e.g., `"1.108.0"` or `"snapshot"`).
 * @returns {TypesConfig} - The package name and version string for the UI5 types.
 */
function getTypes(ui5Version) {
    if (ui5Version?.includes('snapshot')) {
        return {
            typesPackage: "@sapui5/types" /* UI5_DEFAULT.TYPES_PACKAGE_NAME */,
            typesVersion: `~${"1.136.0" /* UI5_DEFAULT.TYPES_VERSION_BEST */}`
        };
    }
    const typesPackage = (0, ui5_config_1.getTypesPackage)(ui5Version);
    const isTypesPackage = typesPackage === "@sapui5/types" /* UI5_DEFAULT.TYPES_PACKAGE_NAME */;
    const typesVersion = isTypesPackage ? (0, ui5_config_1.getTypesVersion)(ui5Version) : (0, ui5_config_1.getEsmTypesVersion)(ui5Version);
    return {
        typesPackage,
        typesVersion
    };
}
/**
 * Get the variant for the CF project.
 *
 * @param {CfAdpWriterConfig} config - The CF configuration.
 * @returns {DescriptorVariant} The variant for the CF project.
 */
function getCfVariant(config) {
    const { app, ui5 } = config;
    const variant = {
        layer: app.layer,
        reference: app.id,
        id: app.namespace,
        namespace: 'apps/' + app.id + '/appVariants/' + app.namespace + '/',
        content: [
            {
                changeType: 'appdescr_ui5_setMinUI5Version',
                content: {
                    minUI5Version: ui5.version
                }
            },
            {
                changeType: 'appdescr_app_setTitle',
                content: {},
                texts: {
                    i18n: 'i18n/i18n.properties'
                }
            }
        ]
    };
    return variant;
}
/**
 * Writes a given project template files within a specified folder in the project directory.
 *
 * @param {string} baseTmplPath - The root path of the templates folder.
 * @param {string} projectPath - The root path of the project.
 * @param {AdpWriterConfig} data - The data to be populated in the template file.
 * @param {Editor} fs - The `mem-fs-editor` instance used for file operations.
 * @returns {void}
 */
function writeTemplateToFolder(baseTmplPath, projectPath, data, fs) {
    const ui5Version = data.ui5?.version;
    const tmplPath = (0, node_path_1.join)(baseTmplPath, 'project', '**/*.*');
    const tsConfigPath = (0, node_path_1.join)(baseTmplPath, 'typescript', 'tsconfig.json');
    const { typesPackage, typesVersion } = getTypes(ui5Version);
    try {
        fs.copyTpl(tmplPath, projectPath, { ...data, typesPackage, typesVersion }, undefined, {
            globOptions: { dot: true },
            processDestinationPath: (filePath) => filePath.replace(/gitignore.tmpl/g, '.gitignore')
        });
        if (data.options?.enableTypeScript) {
            const id = data.app?.id?.split('.').join('/');
            fs.copyTpl(tsConfigPath, (0, node_path_1.join)(projectPath, 'tsconfig.json'), { id, typesPackage }, undefined, {
                globOptions: { dot: true }
            });
        }
    }
    catch (e) {
        throw new Error(`Could not write template files to folder. Reason: ${e.message}`);
    }
}
/**
 * Writes a ui5.yaml file within a specified folder in the project directory.
 *
 * @param {string} projectPath - The root path of the project.
 * @param {AdpWriterConfig} data - The data to be populated in the template file.
 * @param {Editor} fs - The `mem-fs-editor` instance used for file operations.
 * @returns {void}
 */
async function writeUI5Yaml(projectPath, data, fs) {
    try {
        const ui5ConfigPath = (0, node_path_1.join)(projectPath, 'ui5.yaml');
        const baseUi5ConfigContent = fs.read(ui5ConfigPath);
        const ui5Config = await ui5_config_1.UI5Config.newInstance(baseUi5ConfigContent);
        ui5Config.setConfiguration({ propertiesFileSourceEncoding: 'UTF-8' });
        (0, options_1.enhanceUI5YamlWithCustomConfig)(ui5Config, data.customConfig);
        (0, options_1.enhanceUI5YamlWithTranspileMiddleware)(ui5Config, data);
        (0, options_1.enhanceUI5Yaml)(ui5Config, data);
        (0, options_1.enhanceUI5YamlWithCustomTask)(ui5Config, data);
        fs.write(ui5ConfigPath, ui5Config.toString());
    }
    catch (e) {
        throw new Error(`Could not write ui5.yaml file. Reason: ${e.message}`);
    }
}
/**
 * Writes a ui5.yaml file for CF project within a specified folder in the project directory.
 *
 * @param {string} projectPath - The root path of the project.
 * @param {CfAdpWriterConfig} data - The data to be populated in the template file.
 * @param {Editor} fs - The `mem-fs-editor` instance used for file operations.
 * @returns {void}
 */
async function writeCfUI5Yaml(projectPath, data, fs) {
    try {
        const ui5ConfigPath = (0, node_path_1.join)(projectPath, 'ui5.yaml');
        const baseUi5ConfigContent = fs.read(ui5ConfigPath);
        const ui5Config = await ui5_config_1.UI5Config.newInstance(baseUi5ConfigContent);
        ui5Config.setConfiguration({ propertiesFileSourceEncoding: 'UTF-8' });
        (0, options_1.enhanceUI5YamlWithCustomConfig)(ui5Config, data.customConfig);
        /** Builder task */
        (0, options_1.enhanceUI5YamlWithCfCustomTask)(ui5Config, data);
        /** Middlewares */
        (0, options_1.enhanceUI5YamlWithCfCustomMiddleware)(ui5Config, data);
        fs.write(ui5ConfigPath, ui5Config.toString());
    }
    catch (e) {
        throw new Error(`Could not write ui5.yaml file. Reason: ${e.message}`);
    }
}
/**
 * Writes a ui5-deploy.yaml file within a specified folder in the project directory.
 *
 * @param {string} projectPath - The root path of the project.
 * @param {AdpWriterConfig} data - The data to be populated in the template file.
 * @param {Editor} fs - The `mem-fs-editor` instance used for file operations.
 * @returns {void}
 */
async function writeUI5DeployYaml(projectPath, data, fs) {
    try {
        if ((0, options_1.hasDeployConfig)(data)) {
            const ui5ConfigPath = (0, node_path_1.join)(projectPath, 'ui5.yaml');
            const baseUi5ConfigContent = fs.read(ui5ConfigPath);
            const ui5DeployConfig = await ui5_config_1.UI5Config.newInstance(baseUi5ConfigContent);
            (0, options_1.enhanceUI5DeployYaml)(ui5DeployConfig, data);
            fs.write((0, node_path_1.join)(projectPath, 'ui5-deploy.yaml'), ui5DeployConfig.toString());
        }
    }
    catch (e) {
        throw new Error(`Could not write ui5-deploy.yaml file. Reason: ${e.message}`);
    }
}
/**
 * Write CF-specific templates and configuration files.
 *
 * @param {string} basePath - The base path.
 * @param {DescriptorVariant} variant - The descriptor variant.
 * @param {CfAdpWriterConfig} config - The CF configuration.
 * @param {Editor} fs - The memfs editor instance.
 */
async function writeCfTemplates(basePath, variant, config, fs) {
    const baseTmplPath = (0, node_path_1.join)(__dirname, '../../templates');
    const templatePath = config.options?.templatePathOverwrite ?? baseTmplPath;
    const { app, project, options } = config;
    fs.copyTpl((0, node_path_1.join)(templatePath, 'project/webapp/manifest.appdescr_variant'), (0, node_path_1.join)(project.folder, 'webapp', 'manifest.appdescr_variant'), { app: variant });
    fs.copyTpl((0, node_path_1.join)(templatePath, 'cf/package.json'), (0, node_path_1.join)(project.folder, 'package.json'), {
        module: project.name
    });
    fs.copyTpl((0, node_path_1.join)(templatePath, 'cf/ui5.yaml'), (0, node_path_1.join)(project.folder, 'ui5.yaml'), {
        module: project.name
    });
    fs.copyTpl((0, node_path_1.join)(templatePath, 'cf/i18n/i18n.properties'), (0, node_path_1.join)(project.folder, 'webapp/i18n/i18n.properties'), {
        module: project.name,
        moduleTitle: app.title,
        appVariantId: app.namespace,
        i18nGuid: config.app.i18nDescription
    });
    fs.copy((0, node_path_1.join)(templatePath, 'cf/_gitignore'), (0, node_path_1.join)(project.folder, '.gitignore'));
    if (options?.addStandaloneApprouter) {
        fs.copyTpl((0, node_path_1.join)(templatePath, 'cf/approuter/package.json'), (0, node_path_1.join)(basePath, `${project.name}-approuter/package.json`), {
            projectName: project.name
        });
        fs.copyTpl((0, node_path_1.join)(templatePath, 'cf/approuter/xs-app.json'), (0, node_path_1.join)(basePath, `${project.name}-approuter/xs-app.json`), {});
    }
    if (!fs.exists((0, node_path_1.join)(basePath, 'xs-security.json'))) {
        fs.copyTpl((0, node_path_1.join)(templatePath, 'cf/xs-security.json'), (0, node_path_1.join)(basePath, 'xs-security.json'), {
            projectName: project.name
        });
    }
}
