// @sap-ux/store/dist/utils/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getSapToolsDirectory = exports.getFioriToolsDirectory = exports.SapTools = exports.FioriToolsSettings = exports.pick = void 0;
exports.hasAnyValue = hasAnyValue;
exports.errorInstance = errorInstance;
exports.errorString = errorString;
exports.toPersistenceName = toPersistenceName;
exports.getEntityFileName = getEntityFileName;
const node_os_1 = require("node:os");
const node_path_1 = require("node:path");
const pluralize_1 = require("pluralize");
/** Pick the properties listed and return a new object with a shallow-copy */
const pick = (target, ...props) => {
    return ((target &&
        props?.length > 0 &&
        props.reduce((o, k) => {
            o[k] = target[k];
            return o;
        }, {})) ||
        undefined);
};
exports.pick = pick;
/**
 * Checks if any of the values in the object are not `undefined` or `null`
 *
 * @param obj - the object to check
 * @param props - the properties to check on the object
 * @returns - `true` if any value is not `undefined` or `null`, `false` otherwise
 */
function hasAnyValue(obj, props) {
    if (obj == null || typeof obj !== 'object') {
        return false;
    }
    return props.some((prop) => {
        const value = obj[prop];
        return value !== undefined && value !== null;
    });
}
/** Given an `Error` or any other object thrown, returns an `Error` instance */
function errorInstance(e) {
    if (e instanceof Error) {
        return e;
    }
    else {
        return new Error(String(e));
    }
}
/** If input in an instance of `Error` return the message property,
 *  otherwise convert the input to its string representation
 */
function errorString(e) {
    return e instanceof Error ? e.message : String(e);
}
var FioriToolsSettings;
(function (FioriToolsSettings) {
    FioriToolsSettings["dir"] = ".fioritools";
})(FioriToolsSettings || (exports.FioriToolsSettings = FioriToolsSettings = {}));
var SapTools;
(function (SapTools) {
    SapTools["dir"] = ".saptools";
})(SapTools || (exports.SapTools = SapTools = {}));
const getFioriToolsDirectory = () => {
    return (0, node_path_1.join)((0, node_os_1.homedir)(), FioriToolsSettings.dir);
};
exports.getFioriToolsDirectory = getFioriToolsDirectory;
const getSapToolsDirectory = () => {
    return (0, node_path_1.join)((0, node_os_1.homedir)(), SapTools.dir);
};
exports.getSapToolsDirectory = getSapToolsDirectory;
/**
 * Trims, lowercases and returns plural if a non-empty string
 *
 * @param s
 */
function toPersistenceName(s) {
    const t = s?.trim().toLowerCase();
    return t && (0, pluralize_1.plural)(t);
}
function getEntityFileName(entityName) {
    return toPersistenceName(entityName) + '.json';
}
__exportStar(require("./app-studio"), exports);
__exportStar(require("./backend"), exports);
