// @sap-ux/store/dist/data-provider/backend-system.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SystemDataProvider = void 0;
const hybrid_1 = require("../data-access/hybrid");
const backend_system_1 = require("../entities/backend-system");
const constants_1 = require("./constants");
const types_1 = require("../types");
const utils_1 = require("../utils");
const node_path_1 = require("node:path");
const node_fs_1 = require("node:fs");
const SystemDataProvider = class {
    dataAccessor;
    entityName = constants_1.Entities.BackendSystem;
    logger;
    constructor(logger, options = {}) {
        this.logger = logger;
        this.dataAccessor = (0, hybrid_1.getHybridStore)(this.logger, options);
    }
    async read(key) {
        return this.dataAccessor.read({ entityName: this.entityName, id: key.getId() });
    }
    async write(entity) {
        let e;
        if (!(entity instanceof backend_system_1.BackendSystem)) {
            // We need to use the correct class otherwise the annotations are not effective
            e = new backend_system_1.BackendSystem({ ...entity });
        }
        else {
            e = entity;
        }
        return this.dataAccessor.write({
            entityName: this.entityName,
            id: backend_system_1.BackendSystemKey.from(entity).getId(),
            entity: e
        });
    }
    async delete(entity) {
        return this.dataAccessor.del({
            entityName: this.entityName,
            id: backend_system_1.BackendSystemKey.from(entity).getId()
        });
    }
    async getAll(providerRetrievalOptions) {
        const migrationRequired = this.isMigrationRequired();
        const { includeSensitiveData = true, backendSystemFilter } = providerRetrievalOptions ?? {};
        // Always fetch sensitive data if migration is pending
        let systems = await this.dataAccessor.readAll({
            entityName: this.entityName,
            includeSensitiveData: migrationRequired ? true : includeSensitiveData
        });
        const migrationComplete = await this.migrateBackendSystems(systems, includeSensitiveData);
        // If migration occurred, re-read to get the fully updated dataset
        if (!migrationComplete) {
            systems = await this.dataAccessor.readAll({
                entityName: this.entityName,
                includeSensitiveData
            });
        }
        const systemList = Object.values(systems);
        if (!backendSystemFilter) {
            return systemList;
        }
        return this.applyFilters(systemList, backendSystemFilter);
    }
    /**
     * Determines whether a migration is required based on a local marker file.
     */
    isMigrationRequired() {
        const migrationFilePath = (0, node_path_1.join)((0, utils_1.getSapToolsDirectory)(), '.systemsMigrated');
        try {
            const raw = (0, node_fs_1.readFileSync)(migrationFilePath, 'utf-8');
            const data = JSON.parse(raw);
            return !data?.backendSystemMigrationV1;
        }
        catch {
            // If anything fails (file missing, parse error), treat as not migrated
            return true;
        }
    }
    /**
     * Applies filter objects to a list of backend systems.
     */
    applyFilters(systems, filters = {}) {
        return systems.filter((system) => Object.entries(filters).every(([key, value]) => system[key] === value));
    }
    /**
     * Ensures all stored backend systems are migrated to the latest structure.
     */
    async migrateBackendSystems(systems, containsSensitiveData) {
        let allMigrated = true;
        for (const [id, system] of Object.entries(systems)) {
            // Skip already-migrated entries
            const alreadyMigrated = system?.hasSensitiveData !== undefined &&
                system?.systemType !== undefined &&
                system?.connectionType !== undefined;
            if (alreadyMigrated) {
                continue;
            }
            allMigrated = false;
            const migratedSystem = await this.buildMigratedSystem(system, containsSensitiveData, id);
            await this.dataAccessor.partialUpdate({
                entityName: this.entityName,
                id,
                entity: {
                    hasSensitiveData: migratedSystem.hasSensitiveData,
                    systemType: migratedSystem.systemType,
                    connectionType: migratedSystem.connectionType
                }
            });
        }
        if (!allMigrated) {
            // Write migration marker file
            const filePath = (0, node_path_1.join)((0, utils_1.getSapToolsDirectory)(), '.systemsMigrated');
            const marker = { backendSystemMigrationV1: new Date().toISOString() };
            (0, node_fs_1.writeFileSync)(filePath, JSON.stringify(marker, null, 2));
        }
        return allMigrated;
    }
    /**
     * Builds a fully migrated BackendSystem instance.
     */
    async buildMigratedSystem(system, containsSensitiveData, id) {
        let fullSystem = system;
        // Ensure sensitive data available if needed
        if (!containsSensitiveData) {
            fullSystem = await this.dataAccessor.read({
                entityName: this.entityName,
                id
            });
        }
        const inferredSystemType = fullSystem.systemType ?? (0, utils_1.getBackendSystemType)(fullSystem) ?? 'OnPrem';
        const connectionType = fullSystem?.connectionType ?? types_1.ConnectionType.AbapCatalog; // will need to be removed once adding different connection types is possible
        return new backend_system_1.BackendSystem({
            ...fullSystem,
            systemType: inferredSystemType,
            connectionType
        });
    }
};
exports.SystemDataProvider = SystemDataProvider;
