// @sap-ux/preview-middleware/dist/client/flp/init.js
"use strict";

sap.ui.define(["sap/base/Log", "open/ux/preview/client/thirdparty/@sap-ux-private/control-property-editor-common", "sap/ui/core/IconPool", "sap/base/i18n/ResourceBundle", "../adp/api-handler", "../utils/error", "./initCdm", "./initConnectors", "../utils/version", "../utils/info-center-message"], function (Log, ___sap_ux_private_control_property_editor_common, IconPool, ResourceBundle, ___adp_api_handler, ___utils_error, __initCdm, __initConnectors, ___utils_version, ___utils_info_center_message) {
  "use strict";

  function _interopRequireDefault(obj) {
    return obj && obj.__esModule && typeof obj.default !== "undefined" ? obj.default : obj;
  }
  function __ui5_require_async(path) {
    return new Promise(function (resolve, reject) {
      sap.ui.require([path], function (module) {
        if (!(module && module.__esModule)) {
          module = module === null || !(typeof module === "object" && path.endsWith("/library")) ? {
            default: module
          } : module;
          Object.defineProperty(module, "__esModule", {
            value: true
          });
        }
        resolve(module);
      }, function (err) {
        reject(err);
      });
    });
  }
  const MessageBarType = ___sap_ux_private_control_property_editor_common["MessageBarType"];
  const SCENARIO = ___sap_ux_private_control_property_editor_common["SCENARIO"];
  const getManifestAppdescr = ___adp_api_handler["getManifestAppdescr"];
  const getError = ___utils_error["getError"];
  const initCdm = _interopRequireDefault(__initCdm);
  const initConnectors = _interopRequireDefault(__initConnectors);
  const getUi5Version = ___utils_version["getUi5Version"];
  const isLowerThanMinimalUi5Version = ___utils_version["isLowerThanMinimalUi5Version"];
  const sendInfoCenterMessage = ___utils_info_center_message["sendInfoCenterMessage"];
  /**
   * SAPUI5 delivered namespaces from https://ui5.sap.com/#/api/sap
   */
  const UI5_LIBS = ['sap.apf', 'sap.base', 'sap.chart', 'sap.collaboration', 'sap.f', 'sap.fe', 'sap.fileviewer', 'sap.gantt', 'sap.landvisz', 'sap.m', 'sap.ndc', 'sap.ovp', 'sap.rules', 'sap.suite', 'sap.tnt', 'sap.ui', 'sap.uiext', 'sap.ushell', 'sap.uxap', 'sap.viz', 'sap.webanalytics', 'sap.zen'];
  /**
   * Check whether a specific dependency is a custom library, and if yes, add it to the map.
   *
   * @param dependency dependency from the manifest
   * @param customLibs map containing the required custom libraries
   */
  function addKeys(dependency, customLibs) {
    Object.keys(dependency).forEach(function (key) {
      // ignore libs or Components that start with SAPUI5 delivered namespaces
      if (!UI5_LIBS.some(function (substring) {
        return key === substring || key.startsWith(substring + '.');
      })) {
        customLibs[key] = true;
      }
    });
  }

  /**
   * Check whether a specific ComponentUsage is a custom component, and if yes, add it to the map.
   *
   * @param compUsages ComponentUsage from the manifest
   * @param customLibs map containing the required custom libraries
   */
  function getComponentUsageNames(compUsages, customLibs) {
    const compNames = Object.keys(compUsages).map(function (compUsageKey) {
      return compUsages[compUsageKey].name;
    });
    compNames.forEach(function (key) {
      // ignore libs or Components that start with SAPUI5 delivered namespaces
      if (!UI5_LIBS.some(function (substring) {
        return key === substring || key.startsWith(substring + '.');
      })) {
        customLibs[key] = true;
      }
    });
  }

  /**
   * Fetch the manifest for all the given application urls and generate a string containing all required custom library ids.
   *
   * @param appUrls urls pointing to included applications
   * @returns Promise of a comma separated list of all required libraries.
   */
  async function getManifestLibs(appUrls) {
    const result = {};
    const promises = [];
    for (const url of appUrls) {
      promises.push(fetch(`${url}/manifest.json`).then(async resp => {
        const manifest = await resp.json();
        if (manifest) {
          if (manifest['sap.ui5']?.dependencies) {
            if (manifest['sap.ui5'].dependencies.libs) {
              addKeys(manifest['sap.ui5'].dependencies.libs, result);
            }
            if (manifest['sap.ui5'].dependencies.components) {
              addKeys(manifest['sap.ui5'].dependencies.components, result);
            }
          }
          if (manifest['sap.ui5']?.componentUsages) {
            getComponentUsageNames(manifest['sap.ui5'].componentUsages, result);
          }
        }
      }));
    }
    return Promise.all(promises).then(() => Object.keys(result).join(','));
  }

  /**
   * Register the custom libraries and their url with the UI5 loader.
   *
   * @param dataFromAppIndex data returned from the app index service
   */
  function registerModules(dataFromAppIndex) {
    Object.keys(dataFromAppIndex).forEach(function (moduleDefinitionKey) {
      const moduleDefinition = dataFromAppIndex[moduleDefinitionKey];
      if (moduleDefinition?.dependencies) {
        moduleDefinition.dependencies.forEach(function (dependency) {
          if (dependency.url && dependency.url.length > 0 && dependency.type === 'UI5LIB') {
            Log.info('Registering Library ' + dependency.componentId + ' from server ' + dependency.url);
            const compId = dependency.componentId.replace(/\./g, '/');
            const config = {
              paths: {}
            };
            config.paths[compId] = dependency.url;
            sap.ui.loader.config(config);
          }
        });
      }
    });
  }

  /**
   * Fetch the app state from the given application urls, then reset the app state.
   *
   * @param container the UShell container
   */
  async function resetAppState(container) {
    const urlParams = new URLSearchParams(window.location.hash);
    const appStateValue = urlParams.get('sap-iapp-state') ?? urlParams.get('/?sap-iapp-state');
    if (appStateValue) {
      const appStateService = await container.getServiceAsync('AppState');
      appStateService.deleteAppState(appStateValue);
    }
  }

  /**
   * Fetch the manifest from the given application urls, then parse them for custom libs, and finally request their urls.
   *
   * @param appUrls application urls
   * @param urlParams URLSearchParams object
   * @returns returns a promise when the registration is completed.
   */
  async function registerComponentDependencyPaths(appUrls, urlParams) {
    const libs = await getManifestLibs(appUrls);
    if (libs && libs.length > 0) {
      let url = '/sap/bc/ui2/app_index/ui5_app_info?id=' + libs;
      const sapClient = urlParams.get('sap-client');
      if (sapClient?.length === 3) {
        url = url + '&sap-client=' + sapClient;
      }
      const response = await fetch(url);
      try {
        registerModules(await response.json());
      } catch (error) {
        Log.error(`Registering of reuse libs failed. Error:${error}`);
      }
    }
  }

  /**
   * Register SAP fonts that are also registered in a productive Fiori launchpad.
   */
  function registerSAPFonts() {
    //Fiori Theme font family and URI
    const fioriTheme = {
      fontFamily: 'SAP-icons-TNT',
      fontURI: sap.ui.require.toUrl('sap/tnt/themes/base/fonts/')
    };
    //Registering to the icon pool
    IconPool.registerFont(fioriTheme);
    //SAP Business Suite Theme font family and URI
    const suiteTheme = {
      fontFamily: 'BusinessSuiteInAppSymbols',
      fontURI: sap.ui.require.toUrl('sap/ushell/themes/base/fonts/')
    };
    //Registering to the icon pool
    IconPool.registerFont(suiteTheme);
  }

  /**
   * Create Resource Bundle based on the scenario.
   *
   * @param scenario to be used for the resource bundle.
   */
  async function loadI18nResourceBundle(scenario) {
    if (scenario === SCENARIO.AdaptationProject) {
      const manifest = await getManifestAppdescr();
      const enhanceWith = manifest.content.filter(content => content.texts?.i18n).map(content => ({
        bundleUrl: `../${content.texts.i18n}`
      }));
      return ResourceBundle.create({
        url: '../i18n/i18n.properties',
        enhanceWith
      });
    }
    return ResourceBundle.create({
      url: 'i18n/i18n.properties'
    });
  }

  /**
   * Read the application title from the resource bundle and set it as document title.
   *
   * @param resourceBundle resource bundle to read the title from.
   * @param i18nKey optional parameter to define the i18n key to be used for the title.
   */
  function setI18nTitle(resourceBundle, i18nKey = 'appTitle') {
    if (resourceBundle.hasText(i18nKey)) {
      document.title = resourceBundle.getText(i18nKey) ?? document.title;
    }
  }

  /**
   * This function dynamically adds a "Generate Card" action to the SAP Fiori Launchpad for the given component instance.
   *
   * @param componentInstance - The instance of the component for which the card generation action is being added.
   * @param container - The SAP Fiori Launchpad container instance used to access services.
   */
  function addCardGenerationUserAction(componentInstance, container) {
    sap.ui.require(['sap/cards/ap/generator/CardGenerator'], async CardGenerator => {
      const extensionService = await container.getServiceAsync('Extension');
      const controlProperties = {
        icon: 'sap-icon://add',
        id: 'generate_card',
        text: 'Generate Card',
        tooltip: 'Generate Card',
        press: () => {
          CardGenerator.initializeAsync(componentInstance);
        }
      };
      const parameters = {
        controlType: 'sap.ushell.ui.launchpad.ActionItem'
      };
      const generateCardAction = await extensionService.createUserAction(controlProperties, parameters);
      generateCardAction.showForCurrentApp();
    });
  }

  /**
   * Apply additional configuration and initialize sandbox.
   *
   * @param params init parameters read from the script tag
   * @param params.appUrls JSON containing a string array of application urls
   * @param params.flex JSON containing the flex configuration
   * @param params.customInit path to the custom init module to be called
   * @param params.enhancedHomePage boolean indicating if enhanced homepage is enabled
   * @returns promise
   */
  async function init({
    appUrls,
    flex,
    customInit,
    enhancedHomePage,
    enableCardGenerator
  }) {
    // Set CDM configuration before importing ushell container
    // to ensure proper configuration pickup during bootstrap
    if (enhancedHomePage) {
      initCdm();
    }
    const urlParams = new URLSearchParams(window.location.search);
    const container = sap?.ushell?.Container ?? (await __ui5_require_async('sap/ushell/Container')).default;
    let scenario = '';
    const ui5VersionInfo = await getUi5Version();
    // Register RTA if configured
    if (flex) {
      const flexSettings = JSON.parse(flex);
      scenario = flexSettings.scenario;
      container.attachRendererCreatedEvent(async function () {
        const lifecycleService = await container.getServiceAsync('AppLifeCycle');
        lifecycleService.attachAppLoaded(event => {
          const view = event.getParameter('componentInstance');
          const pluginScript = flexSettings.pluginScript ?? '';
          const libs = [];
          if (isLowerThanMinimalUi5Version(ui5VersionInfo, {
            major: 1,
            minor: 72
          })) {
            libs.push('open/ux/preview/client/flp/initRta');
          } else {
            libs.push('sap/ui/rta/api/startAdaptation');
          }
          if (flexSettings.pluginScript) {
            libs.push(pluginScript);
            delete flexSettings.pluginScript;
          }
          const options = {
            rootControl: view,
            validateAppVersion: false,
            flexSettings
          };
          sap.ui.require(libs, async function (startAdaptation, pluginScript) {
            try {
              await startAdaptation(options, pluginScript);
            } catch (error) {
              await sendInfoCenterMessage({
                title: {
                  key: 'FLP_ADAPTATION_START_FAILED_TITLE'
                },
                description: getError(error).message,
                type: MessageBarType.error
              });
              await handleHigherLayerChanges(error, ui5VersionInfo);
            }
          });
        });
      });
    }
    if (enableCardGenerator && !isLowerThanMinimalUi5Version(ui5VersionInfo, {
      major: 1,
      minor: 121
    })) {
      container.attachRendererCreatedEvent(async function () {
        const lifecycleService = await container.getServiceAsync('AppLifeCycle');
        lifecycleService.attachAppLoaded(event => {
          const componentInstance = event.getParameter('componentInstance');
          addCardGenerationUserAction(componentInstance, container);
        });
      });
    } else {
      Log.warning('Card generator is not supported for the current UI5 version.');
    }

    // reset app state if requested
    if (urlParams.get('fiori-tools-iapp-state')?.toLocaleLowerCase() !== 'true') {
      await resetAppState(container);
    }

    // Load custom library paths if configured
    if (appUrls) {
      await registerComponentDependencyPaths(JSON.parse(appUrls) ?? [], urlParams);
    }

    // Load rta connector
    await initConnectors();

    // Load custom initialization module
    if (customInit) {
      sap.ui.require([customInit]);
    }

    // init
    const resourceBundle = await loadI18nResourceBundle(scenario);
    setI18nTitle(resourceBundle);
    registerSAPFonts();
    if (enhancedHomePage) {
      await container.init('cdm');
    }
    const renderer = ui5VersionInfo.major < 2 && !ui5VersionInfo.label?.includes('legacy-free') ? await container.createRenderer(undefined, true) : await container.createRendererInternal(undefined, true);
    renderer.placeAt('content');
  }

  // eslint-disable-next-line @sap-ux/fiori-tools/sap-no-dom-access,@sap-ux/fiori-tools/sap-browser-api-warning
  const bootstrapConfig = document.getElementById('sap-ui-bootstrap');
  if (bootstrapConfig) {
    init({
      appUrls: bootstrapConfig.dataset.openUxPreviewLibsManifests,
      flex: bootstrapConfig.dataset.openUxPreviewFlexSettings,
      customInit: bootstrapConfig.dataset.openUxPreviewCustomInit,
      enhancedHomePage: !!bootstrapConfig.dataset.openUxPreviewEnhancedHomepage,
      enableCardGenerator: !!bootstrapConfig.dataset.openUxPreviewEnableCardGenerator
    }).catch(e => {
      const error = getError(e);
      Log.error('Sandbox initialization failed: ' + error.message);
    });
  }

  /**
   * Handle higher layer changes when starting UI Adaptation.
   * When RTA detects higher layer changes an error with Reload triggered text is thrown, the RTA instance is destroyed and the application is reloaded.
   * For UI5 version lower than 1.84.0 RTA is showing a popup with notification text about the detection of higher layer changes.
   *
   * @param error the error thrown when there are higher layer changes when starting UI Adaptation.
   * @param ui5VersionInfo ui5 version info
   */
  async function handleHigherLayerChanges(error, ui5VersionInfo) {
    const err = getError(error);
    if (err.message.includes('Reload triggered')) {
      if (!isLowerThanMinimalUi5Version(ui5VersionInfo, {
        major: 1,
        minor: 84
      })) {
        await sendInfoCenterMessage({
          title: {
            key: 'HIGHER_LAYER_CHANGES_TITLE'
          },
          description: {
            key: 'HIGHER_LAYER_CHANGES_INFO_MESSAGE'
          },
          type: MessageBarType.warning
        });
      }

      // eslint-disable-next-line @sap-ux/fiori-tools/sap-no-location-reload
      window.location.reload();
    }
  }
  var __exports = {
    __esModule: true
  };
  __exports.resetAppState = resetAppState;
  __exports.registerComponentDependencyPaths = registerComponentDependencyPaths;
  __exports.registerSAPFonts = registerSAPFonts;
  __exports.loadI18nResourceBundle = loadI18nResourceBundle;
  __exports.setI18nTitle = setI18nTitle;
  __exports.init = init;
  __exports.handleHigherLayerChanges = handleHigherLayerChanges;
  return __exports;
});
