// @sap-ux/adp-tooling/dist/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.cfServicesPromptNames = exports.AppRouterType = exports.ChangeTypeMap = exports.ApplicationType = void 0;
var ApplicationType;
(function (ApplicationType) {
    ApplicationType["FIORI_ELEMENTS"] = "FioriElements";
    ApplicationType["FIORI_ELEMENTS_OVP"] = "FioriElementsOVP";
    ApplicationType["FREE_STYLE"] = "FreeStyle";
    ApplicationType["NONE"] = "";
})(ApplicationType || (exports.ApplicationType = ApplicationType = {}));
/**
 * A mapping of ChangeType values to their respective change names.
 */
exports.ChangeTypeMap = {
    ["appdescr_ui5_addNewModel" /* ChangeType.ADD_NEW_MODEL */]: 'addNewModel',
    ["appdescr_app_addAnnotationsToOData" /* ChangeType.ADD_ANNOTATIONS_TO_ODATA */]: 'addAnnotationsToOData',
    ["appdescr_app_changeDataSource" /* ChangeType.CHANGE_DATA_SOURCE */]: 'changeDataSource',
    ["appdescr_ui5_addComponentUsages" /* ChangeType.ADD_COMPONENT_USAGES */]: 'addComponentUsages',
    ["appdescr_ui5_addLibraries" /* ChangeType.ADD_LIBRARY_REFERENCE */]: 'addLibraries',
    ["appdescr_app_changeInbound" /* ChangeType.CHANGE_INBOUND */]: 'changeInbound'
};
exports.AppRouterType = {
    MANAGED: 'Managed HTML5 Application Runtime',
    STANDALONE: 'Standalone HTML5 Application Runtime'
};
/**
 * CF services (application sources) prompts
 */
var cfServicesPromptNames;
(function (cfServicesPromptNames) {
    cfServicesPromptNames["approuter"] = "approuter";
    cfServicesPromptNames["businessService"] = "businessService";
    cfServicesPromptNames["businessSolutionName"] = "businessSolutionName";
    cfServicesPromptNames["baseApp"] = "baseApp";
})(cfServicesPromptNames || (exports.cfServicesPromptNames = cfServicesPromptNames = {}));
